### ISSUE TITLE
find() query on a model by non existing property returns whole collection

### ISSUE THREAD

**Issue by Kuzzy** [Automattic/mongoose#10763]
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

find() query on a model by non existing property returns whole collection.

**If the current behavior is a bug, please provide the steps to reproduce.**

```javascript
const templateSchema = mongoose.Schema({
  title: { type: String, required: true },
}, {
  timestamps: true,
  versionKey: false,
})

const Template = mongoose.model('Template', templateSchema)

await Template.create({ title: 'some title' })
const queryResult = await Template.find({ test: ' test' })
console.log(queryResult)
// [{ 
// _id: new ObjectId("5adf374309ee5958d92c4133"),
//  updatedAt: 2020-03-19T12:50:33.280Z,
//  createdAt: 2020-04-24T13:55:15.779Z,
//  title: 'some title',
// }]
```

If enable mongoose debug, next query to DB will be logged:
```
Mongoose: templates.find({}, { projection: {} })
```

**What is the expected behavior?**

1. find() query on a model by non existing property returns empty array. 
2. Not sure here, query to DB should be templates.find({ test: 'test' }, { projection: {} }) or DB should not be queried at all?

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**
Node.js v14.17.4
Mongoose: 6.0.6

**Comment by Kamikadze4GAME** [comment]
Hello @Kuzzy 

This is expected behavior. Please read [migration to 6 guide](https://mongoosejs.com/docs/migrating_to_6.html#strictquery-is-removed-and-replaced-by-strict)

```javascript
const userSchema = new Schema({ name: String });
const User = mongoose.model('User', userSchema);

// By default, this is equivalent to `User.find()` because Mongoose filters out `notInSchema`
await User.find({ notInSchema: 1 });
```

**Comment by Kuzzy** [comment]
Thanks @Kamikadze4GAME, thats it.

But is there a way to apply this option to each query? Because it's to risky to have such behaviour by default, one wrong query may overload app memory and crash it...

**Comment by Kuzzy** [comment]
@Kamikadze4GAME yes, but this global setting will also let to save `notInSchema` values when creating or updating document. I want to override this behaviour only for `find` query method and I've found solution. I've created small plugin:

```javascript
// MongooseStrictFindPlugin.js
module.exports = function MongooseStrictFindPlugin(schema) {
  schema.pre('find', function (next) {
    this.setOptions({ strict: false })
    next()
  })
}

// index.js
mongoose.plugin(MongooseStrictFindPlugin)
```

Now it works like expected, but is there any caveats for such approach?

**Comment by Kuzzy** [comment]
Set `strict: false` option can't be a solution because` find` query will return whole document do not despite on which fields declared in scheme. I think this is wrong behaviour. Why mongoose modifying query in such way? MongoDB will never return whole collection if you pass in query non existing property. Small typo error in code may cause a big problems in production env. So I will need to implement additional layer which will validate each query object through schema's fields and delete not existing. It's not very user friendly :(

**Comment by xoxoxo** [comment]
I was pointed here and I agree this should not behave like it is now. It should reflect mongodb. It could easily lead to some security issues or just dumb errors. I want my schema to be validated on update/save only.  Here is example that could lead to sec issues.

```
const userSchema = mongoose.Schema({
  userName: { type: String, required: true },
})
```

```
userSchema.findOne({
  usersName:  'someuser'
});
```

It will return random user from db as we made a mistake and added extra S in 'users';

**Comment by ilaydatosun** [comment]
@vkarpov15 I think your solution is not a very good option. So adding the ('strictQuery', false) option means going around the whole database. It is necessary to specify the field that you want to search in the Schema definition. For example, if you want to search query.find({age:{$lt:23}}), **{age:Number}** should be specified in the Schema definition.

**Comment by florianwalther-private** [comment]
I just wanna chime in and agree that this shouldn't be the default behavior and it can easily cause some devastating bugs/security holes with just a little typo in the query object key.

But I'm happy that we can disable it globally!

**Comment by gierschv** [comment]
> I just wanna chime in and agree that this shouldn't be the default behavior and it can easily cause some devastating bugs/security holes with just a little typo in the query object key.
> 
> 
> 
> But I'm happy that we can disable it globally!

+1, also I think this "strict" mode has the wrong behaviour, it should throw/reject queries instead of letting them run by modifying them.

Another example, let's say you run a deleteMany/deleteOne with only one property that has a typo, you san empty your collection or randomly delete any document in it. If you run the same query in your monogosh, nothing critical will happened.
